### ISSUE TITLE
[Enhancement] Data libraries point to S3 location

### ISSUE THREAD

**Issue by MatthewRalston** [galaxyproject/bioblend#177]
Hi everyone,
I have an idea for an enhancement to both Galaxy and bioblend where data is hosted remotely on an S3 account that the Galaxy admin has access to. My understanding of the current "upload_file_from_url" method is that the data is fetched from the remote location and stored locally. You can imagine situations where data is stored in an S3 bucket for low-cost storage, but data shouldn't be retrieved until it is needed (e.g. high cost storage on EBS in comparison to S3.)

**Comment by afgane** [comment]
This sounds very much like Galaxy's [ObjectStore S3 module](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/objectstore/s3.py), which will use S3 as the permanent data repository while caching the working data on a local disc. Not having looked at the object store implementation for a long while, I'm not sure if it applies to data libraries as well but I would guess it will. As a word of caution, the S3 module has seen very little actual use so not sure about it's robustness  but it is very high on the list of features that will see some attention in the coming months.

**Comment by MatthewRalston** [comment]
You're right, it does seem similar to this. I was under the impression that ObjectStore only applies to intermediate datasets. @dannon or @natefoo am I looking simply for ObjectStore? Perhaps there only needs to be an appropriate method in bioblend and/or the Galaxy API so that Galaxy knows to treat the S3 file location like a "linked" file, and actually acquire the file when it is needed? Something like `bioblend.libraries.link_from_S3(bucket_name, file_path)` for uploading?

**Comment by nsoranzo** [comment]
See https://github.com/galaxyproject/galaxy/pull/12533 for the implementation of deferred datasets in Galaxy, which is being released in v22.05 . This has some new API endpoints that BioBlend could be extended to support.
